### PR TITLE
fix dropstored for sparse matrices

### LINKED ISSUE #20513: dropstored! had one job
Tracker for https://github.com/JuliaLang/julia/pull/20488#discussion_r99968442. `dropstored!` appears broken:
```julia
julia> foo = sprand(4, 4, 0.5)
4×4 sparse matrix with 10 Float64 stored entries:
  [2, 1]  =  0.986522
  [1, 2]  =  0.331456
  [2, 2]  =  0.181261
  [4, 2]  =  0.545928
  [1, 3]  =  0.864019
  [2, 3]  =  0.48957
  [3, 3]  =  0.575748
  [4, 3]  =  0.509864
  [1, 4]  =  0.368723
  [3, 4]  =  0.332495

julia> Base.SparseArrays.dropstored!(foo, 2, 1)
4×4 sparse matrix with 9 Float64 stored entries:Error showing value of type SparseMatrixCSC{Float64,Int64}:
ERROR: BoundsError: attempt to access 9-element Array{Int64,1} at index [0]
Stacktrace:
 [1] show(::IOContext{Base.Terminals.TTYTerminal}, ::SparseMatrixCSC{Float64,Int64}) at ./sparse/sparsematrix.jl:147
 [2] show(::IOContext{Base.Terminals.TTYTerminal}, ::MIME{Symbol("text/plain")}, ::SparseMatrixCSC{Float64,Int64}) at ./sparse/sparsematrix.jl:122
 [3] display(::Base.REPL.REPLDisplay{Base.REPL.LineEditREPL}, ::MIME{Symbol("text/plain")}, ::SparseMatrixCSC{Float64,Int64}) at ./REPL.jl:122
 [4] display(::Base.REPL.REPLDisplay{Base.REPL.LineEditREPL}, ::SparseMatrixCSC{Float64,Int64}) at ./REPL.jl:125
 [5] display(::SparseMatrixCSC{Float64,Int64}) at ./multimedia.jl:194

julia> foo.colptr
5-element Array{Int64,1}:
  0
  1
  4
  8
 10
```
Best!

### PR DESCRIPTION
**Pull request by KristofferC**
Should fix #20513

### CODE REVIEW

**Review by tkelman — APPROVED** [review]
who's counting anyway

**Comment by KristofferC** [comment]
I did not understand your comment.

**Comment by StefanKarpinski** [comment]
Would it make sense for `dropstored!` to be a method of `delete!`?

**Comment by KristofferC** [comment]
Probably no. I would argue they have different semantics.

**Comment by StefanKarpinski** [comment]
Don't they both delete the value associated with keys, if there is one, reverting back to a default value, and returning the collection?

File: base/sparse/sparsematrix.jl [hunk 1]
```diff
@@ -2745,7 +2745,7 @@ function dropstored!(A::SparseMatrixCSC, i::Integer, j::Integer)
         # Entry A[i,j] is stored. Drop and return.
         deleteat!(A.rowval, searchk)
         deleteat!(A.nzval, searchk)
-        @simd for m in j:(A.n + 1)
+        @simd for m in (j+1):(A.n + 1)
```
Comment (Sacha0) [inline base/sparse/sparsematrix.jl]: Good catch!

**Review by Sacha0 — APPROVED** [review]
Thanks for fixing this! :)

**Comment by KristofferC** [comment]
@StefanKarpinski  for example:

```julia
julia> d = Dict(1 => 1.0);

julia> delete!(d, 1);

julia> d[1]
ERROR: KeyError: key 1 not found
Stacktrace:
 [1] getindex(::Dict{Int64,Float64}, ::Int64) at ./dict.jl:474

julia> x = sparse([1]);

julia> Base.SparseArrays.dropstored!(x, 1);

julia> x[1]
0
```

**Comment by StefanKarpinski** [comment]
```jl
julia> using DataStructures

julia> d = DefaultDict(0.0, 1 => 1.0)
DataStructures.DefaultDict{Int64,Float64,Float64} with 1 entry:
  1 => 1.0

julia> delete!(d, 1)
DataStructures.DefaultDict{Int64,Float64,Float64} with 0 entries

julia> d[1]
0.0
```

**Comment by StefanKarpinski** [comment]
I.e. the difference is only because the standard `Dict` has no default value while sparse matrices and vectors do. The generic `delete!` operation covers both of these behaviors.

**Comment by Sacha0** [comment]
> Would it make sense for dropstored! to be a method of delete!?

+1 for this suggestion. @KristofferC, have you other semantic differences in mind? Best!

**Comment by KristofferC** [comment]
Nah, I was wrong :)

**Comment by tkelman** [comment]
https://github.com/JuliaLang/julia/pull/20531 would make it not such a good idea though, as deleteat! resizes but dropstored! does not.

**Comment by StefanKarpinski** [comment]
That seems like a somewhat gotcha distinction that should be eliminated.

**Comment by tkelman** [comment]
they serve different purposes, and resizing deleteat! doesn't work for anything other than 1D

**Comment by StefanKarpinski** [comment]
Then it seems mostly like `delete!` and `deleteat!` should not be merged after all.

**Comment by tkelman** [comment]
Depends whether people expect delete! to resize or not. Is that only defined for Associatives right now? In that case resizing isn't all that meaningful. I could see these 3 functions being consolidated into 2 in a few combinations, but not into 1.

**Comment by StefanKarpinski** [comment]
The thinking behind `deleteat!` being separate from `delete!` is coming back to me and IIRC, it was precisely that `delete!` didn't make sense in an indexed collection because it has the effect of change all subsequent "keys" – i.e. indices. The `delete!` function, on the other hand only affects the key you ask to delete, which applies to sparse matrices/vectors and dicts, but not arrays.

**Pull request merged**